### PR TITLE
[merged] Fetcher deallocation cleanups

### PR DESCRIPTION
**Pull request by cgwalters**

### CODE REVIEW

**Comment by mbarnes** [comment]
Looking at your stack trace in https://github.com/ostreedev/ostree/issues/373, it appears the session thread's `GMainContext` is somehow getting freed prematurely (based on `context=0xa0a0a0a0a0a0a0a0`). `OstreeFetcher` really doesn't do anything with it directly so I don't see any error there.

The test code looks good and works for me but the `finalize()` changes strike me as unnecessary on first impression since the shared structs are all ref-counted.  Thinking this could be a ref-counting bug but I'm still trying to reproduce it...

**Comment by cgwalters** [comment]
Currently thinking this is a libsoup bug where in `idle_run_queue()` if we hit `priv->disposed` the source will be destroyed by the main loop, but left in `priv->run_queue_sources`.  Adding debug prints now to see.

**Comment by mbarnes** [comment]
In addition to https://bugzilla.gnome.org/show_bug.cgi?id=768567, it might help to ensure the `SoupSession` is disposed in the same thread where it was created.  Otherwise there may yet be race conditions at play here.
```
diff --git a/src/libostree/ostree-fetcher.c b/src/libostree/ostree-fetcher.c
index 313df6a..10f361a 100644
--- a/src/libostree/ostree-fetcher.c
+++ b/src/libostree/ostree-fetcher.c
@@ -44,7 +44,7 @@ typedef enum {
 typedef struct {
   volatile int ref_count;
 
-  SoupSession *session;
+  SoupSession *session;  /* not referenced */
   GMainContext *main_context;
   GMainLoop *main_loop;
 
@@ -140,7 +140,8 @@ thread_closure_unref (ThreadClosure *thread_closure)
 
   if (g_atomic_int_dec_and_test (&thread_closure->ref_count))
     {
-      g_clear_object (&thread_closure->session);
+      /* The session thread should have cleared this by now. */
+      g_warn_if_fail (thread_closure->session == NULL);
 
       g_clear_pointer (&thread_closure->main_context, g_main_context_unref);
       g_clear_pointer (&thread_closure->main_loop, g_main_loop_unref);
@@ -459,6 +460,7 @@ ostree_fetcher_session_thread (gpointer data)
    * context for this thread before creating the session. */
   g_main_context_push_thread_default (closure->main_context);
 
+  /* We retain ownership of the SoupSession reference. */
   closure->session = soup_session_async_new_with_options (SOUP_SESSION_USER_AGENT, "ostree ",
                                                           SOUP_SESSION_SSL_USE_SYSTEM_CA_FILE, TRUE,
                                                           SOUP_SESSION_USE_THREAD_CONTEXT, TRUE,
@@ -481,6 +483,11 @@ ostree_fetcher_session_thread (gpointer data)
 
   g_main_loop_run (closure->main_loop);
 
+  /* Since the ThreadClosure may be finalized from any thread we
+   * unreference the SoupSession ourself to ensure it's freed in
+   * the same thread where it was created. */
+  g_clear_pointer (&closure->session, g_object_unref);
+
   g_main_context_pop_thread_default (closure->main_context);
 
   thread_closure_unref (closure);
```

**Comment by cgwalters** [comment]
Ok, I folded that change (with a bit more added) into the series and cleaned things up more.

**Comment by cgwalters** [comment]
So far, this appears to fix things *without* the libsoup patch in https://bugzilla.gnome.org/show_bug.cgi?id=768567 which is a good sign too.

File: tests/libtest.sh [hunk 1]
```diff
@@ -138,6 +138,7 @@ assert_file_has_content () {
     if ! grep -q -e "$2" "$1"; then
         sed -e 's/^/# /' < "$1" >&2
         echo 1>&2 "File '$1' doesn't match regexp '$2'"
+	cat $1
```
Comment (mbarnes) [inline tests/libtest.sh]: Is this a debugging leftover?

**Comment by mbarnes** [comment]
Looks pretty bullet-proof to me.   :+1: with warm fuzzies.

Note the `libtest.sh` nit though.

**Comment by mbarnes** [comment]
Something else you might want to sneak in here is the `GTask` in the `OstreeFetcherPendingURI` struct is unused.

File: tests/libtest.sh (hunk 1 above)
Comment (cgwalters) [inline tests/libtest.sh]: Were you reading the "squashed" diff?  I split this into its own commit in e39f4188f644fba77888433a4b93c0da70a33eb3

**Comment by mbarnes** [comment]
:+1:

**Comment by mbarnes** [comment]
@rh-atomic-bot r+
